utilities/rate_limiters/write_amp_based_rate_limiter.cc

```diff
@@ -0,0 +1,466 @@
+      normal_pace_up_(false),
+      percent_delta_(0) {
+  std::fill(total_requests_, total_requests_ + Env::IO_TOTAL, 0);
+  std::fill(total_bytes_through_, total_bytes_through_ + Env::IO_TOTAL, 0);
+}
+
+WriteAmpBasedRateLimiter::~WriteAmpBasedRateLimiter() {
+  MutexLock g(&request_mutex_);
+  stop_ = true;
+  for (auto i = 0; i < Env::IO_TOTAL; ++i) {
+    requests_to_wait_ += queue_[i].size();
+    for (auto& r : queue_[i]) {
+      r->cv.Signal();
+    }
+  }
+  while (requests_to_wait_ > 0) {
+    exit_cv_.Wait();
+  }
+}
+
+void WriteAmpBasedRateLimiter::SetBytesPerSecond(int64_t bytes_per_second) {
+  assert(bytes_per_second > 0);
+  if (auto_tuned_.load(std::memory_order_acquire)) {
+    max_bytes_per_sec_.store(bytes_per_second, std::memory_order_relaxed);
+  } else {
+    SetActualBytesPerSecond(bytes_per_second);
+  }
+}
+
+void WriteAmpBasedRateLimiter::SetAutoTuned(bool auto_tuned) {
+  MutexLock g(&auto_tuned_mutex_);
+  if (auto_tuned_.load(std::memory_order_acquire) != auto_tuned) {
+    if (auto_tuned) {
+      max_bytes_per_sec_.store(rate_bytes_per_sec_, std::memory_order_relaxed);
+      refill_bytes_per_period_.store(
+          CalculateRefillBytesPerPeriod(rate_bytes_per_sec_),
+          std::memory_order_relaxed);
+    } else {
+      // must hold this lock to avoid tuner changing `rate_bytes_per_sec_`
+      MutexLock g2(&request_mutex_);
+      rate_bytes_per_sec_ = max_bytes_per_sec_.load(std::memory_order_relaxed);
+      refill_bytes_per_period_.store(
+          CalculateRefillBytesPerPeriod(rate_bytes_per_sec_),
+          std::memory_order_relaxed);
+    }
+    auto_tuned_.store(auto_tuned, std::memory_order_release);
+  }
+}
+
+void WriteAmpBasedRateLimiter::SetActualBytesPerSecond(
+    int64_t bytes_per_second) {
+  rate_bytes_per_sec_ = bytes_per_second;
+  refill_bytes_per_period_.store(
+      CalculateRefillBytesPerPeriod(bytes_per_second),
+      std::memory_order_relaxed);
+}
+
+bool WriteAmpBasedRateLimiter::IsFrontOfOneQueue(Req* r) {
+  request_mutex_.AssertHeld();
+  for (auto i = 0; i < Env::IO_TOTAL; ++i) {
+    if (!queue_[i].empty() && r == queue_[i].front()) {
+      return true;
+    }
+  }
+  return false;
+}
+
+void WriteAmpBasedRateLimiter::Request(int64_t bytes, const Env::IOPriority pri,
+                                       Statistics* stats) {
+  TEST_SYNC_POINT("WriteAmpBasedRateLimiter::Request");
+  TEST_SYNC_POINT_CALLBACK("WriteAmpBasedRateLimiter::Request:1",
+                           &rate_bytes_per_sec_);
+  if (auto_tuned_.load(std::memory_order_acquire) &&
+      (pri == Env::IO_HIGH || pri == Env::IO_USER) &&
+      duration_highpri_bytes_through_ + duration_bytes_through_ + bytes <=
+          max_bytes_per_sec_.load(std::memory_order_relaxed) * secs_per_tune_) {
+    // In the case where low-priority request is absent, actual time elapsed
+    // will be larger than secs_per_tune_, making the limit even tighter.
+    total_bytes_through_[pri] += bytes;
```

Comment:
Does not this need to be protected by the mutex?

Reply:
I think it should, although the original code does not. Good catch. Moving this part below inside the mutex guard scope. I evaluated lock-free atomic operations a little bit, but since there are many fields need to be refactored that way, the overhead would pretty much comparable with holding a mutex,  and this is the write path, lock contention shouldn't be a major concern, so I think it is fine to just use the mutex to protect it.

Reply:
So we may also need a fix for the original code.